### PR TITLE
Implement offline backup

### LINKED ISSUE #225: Feature: Support offline consistent backups
**User Story**
As a user, I want to take consistent backup of my workloads even if that means I have to take it offline. Examples are: nightly back up of database of Confluence, sentry etc. Some of these applications are deployed via Helm. So, user has no control over how the DB is provisioned.

### PR DESCRIPTION
**Pull request by diptadas**
Fixes #225 #216

Changes:
- Rename package `schedule` to `backup` and add `offline` flag
- Use init container for offline backup instead of sidecar
- Add `check` command
- Create check job from init container
- Create cronjob from restic watcher to delete workload pods, set restic as cronjob's owner ref, set label `app=stash`, set annotation `restic-name`, `stash-operation`
- Update cronjob accordingly when restic updated
- Watch job with label `app=stash`, delete completed jobs/pods for backup/check/recovery

Issues:
- Event writing
- Wait for workload ready in `kutil`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -16,7 +17,9 @@ func main() {
 	defer logs.FlushLogs()
 
 	if err := cmds.NewCmdStash(Version).Execute(); err != nil {
+		log.Infoln("Error in Stash Main:", err)
```
Comment (tamalsaha) [inline main.go]: Use log.Fatalln() instead of 2 commands

File: pkg/check/check.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/record"
+)
+
+type Controller struct {
+	k8sClient   kubernetes.Interface
+	stashClient cs.StashV1alpha1Interface
+	namespace   string
```
Comment (tamalsaha) [inline pkg/check/check.go]: Create an Options struct for the string variables

File: pkg/check/check.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	}
+}
+
+func (c *Controller) Run() (err error) {
+	restic, err := c.stashClient.Restics(c.namespace).Get(c.resticName, metav1.GetOptions{})
+	if err != nil {
+		return
+	}
+
+	defer func() {
```
Comment (tamalsaha) [inline pkg/check/check.go]: Where is the delay?

File: pkg/cmds/backup.go [hunk 4]
```diff
@@ -61,22 +60,22 @@ func NewCmdSchedule() *cobra.Command {
...
-			defer close(stopBackup)
+			ctrl := backup.New(kubeClient, stashClient, opt)
 
-			// split code from here for leader election
-			switch opt.Workload.Kind {
-			case api.KindDeployment, api.KindReplicaSet, api.KindReplicationController:
-				ctrl.ElectLeader(stopBackup)
-			default:
-				ctrl.SetupAndRun(stopBackup)
+			if opt.RunOffline {
```
Comment (tamalsaha) [inline pkg/cmds/backup.go]: Use an option called `runOnce`

File: pkg/controller/jobs.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	"k8s.io/apimachinery/pkg/watch"
+	batch_listers "k8s.io/client-go/listers/batch/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+func (c *StashController) initJobWatcher() {
+	lw := &cache.ListWatch{ // TODO @ Dipta: only watch stash jobs
+		ListFunc: func(options metav1.ListOptions) (rt.Object, error) {
+			return c.k8sClient.BatchV1().Jobs(core.NamespaceAll).List(options)
```
Comment (tamalsaha) [inline pkg/controller/jobs.go]: See an example here: https://github.com/appscode/swanc/blob/67cf9730691c93d92dd363f34d4b16c8147e4e0c/lib/nodes.go#L20

File: pkg/controller/jobs.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		return err
+	}
+	if !exists {
+		fmt.Printf("Job %s does not exist anymore\n", key)
+		return nil
+	} else {
+		job := obj.(*batch.Job)
+		fmt.Printf("Sync/Add/Update for Job %s\n", job.GetName())
+
+		if job.Labels["app"] == util.AppLabelStash && job.Status.Succeeded > 0 {
```
Comment (tamalsaha) [inline pkg/controller/jobs.go]: These should be annotations

File: pkg/controller/restics.go [hunk 7]
```diff
@@ -165,6 +165,27 @@ func (c *StashController) runResticInjector(key string) error {
 		d := obj.(*api.Restic)
 		fmt.Printf("Sync/Add/Update for Restic %s\n", d.GetName())
 
+		if d.Spec.Type == api.BackupOffline {
```
Comment (tamalsaha) [inline pkg/controller/restics.go]: I think this should be done at the start of operator and we should check that the image exists and exit if the image is missing. To check image existence use,
https://github.com/k8sdb/apimachinery/blob/3f92c150ed3bedd0d2d19bcc3856360937d06033/pkg/docker/checks.go#L13

File: pkg/util/kubernetes.go [hunk 8]
```diff
@@ -100,10 +115,19 @@ func WaitUntilSidecarAdded(kubeClient kubernetes.Interface, namespace string, se
 		var podsToRestart []core.Pod
 		for _, pod := range pods.Items {
 			found := false
-			for _, c := range pod.Spec.Containers {
-				if c.Name == StashContainer {
-					found = true
-					break
+			if offline {
```
Comment (tamalsaha) [inline pkg/util/kubernetes.go]: Instead of `offline bool`, I think we should pass the `backupType`

File: pkg/util/kubernetes.go [hunk 9]
```diff
@@ -161,6 +194,16 @@ func GetString(m map[string]string, key string) string {
...
+	container := CreateSidecarContainer(r, tag, workload)
+	container.Args = append(container.Args, "--offline=true")
+	container.Args = append(container.Args, "--tag="+tag)
+	if rbac {
+		container.Args = append(container.Args, "--rbac=true")
+	}
+	return container
+}
+
 func CreateSidecarContainer(r *api.Restic, tag string, workload api.LocalTypedReference) core.Container {
```
Comment (tamalsaha) [inline pkg/util/kubernetes.go]: Where is serviceAccountName created

File: pkg/backup/scheduler.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	if kerr.IsNotFound(err) {
+		err = nil
+		return
+	} else if err != nil {
+		return
+	}
+
+	err = c.resticCLI.Check()
+	if err != nil {
+		c.recorder.Eventf(resource.ObjectReference(), core.EventTypeWarning, eventer.EventReasonFailedToCheck, "Repository check failed for workload %s %s/%s. Reason: %v", c.opt.Workload.Kind, c.opt.Namespace, c.opt.Workload.Name, err)
```
Comment (tamalsaha) [inline pkg/backup/scheduler.go]: When the line is this long, move each parameter to a separate line. Makes it more readable.

File: pkg/util/kubernetes.go (hunk 9 above)
Comment (diptadas) [reply pkg/util/kubernetes.go]: InitContainer and Sidecar use the same service account as the workload?

**Pull request merged**